Add basic tests for HdMaya / mtoh

pmolodo opened the pull request:
I've ported over the tests from the [old maya-to-hydra repo](https://github.com/LumaPictures/maya-to-hydra). They don't really cover much, but they at least provide a framework to add more.

They include some tools for doing basic image diffs to maya snapshots, which could be generally useful.

The first 3 commits are more general test fixes, and could be broken off into a separate PR if folks think this one is too big...

kxl-adsk commented:
@mattyjams I thought it will be useful for you to review this change since you were already adding interactive tests and was missing the ability to diff images. Such utility is provided in this PR and I would like to make sure it could work for you too.

@HamedSabri-adsk please have a look at cmake changes.

mattyjams reviewed (approved):
Looks good to me! Just one question as to whether we can assume Embree is available though. We can maybe gate those checks on a query to `Plug` to see if it is.

The image diffing stuff looks good too! Would be great to get the existing `pxrUsdMayaGL` tests setup to use that.

HamedSabri-adsk reviewed (changes requested):
Hi @pmolodo,  all 4 tests fail for me on Windows. I ran these test against Maya PR 120 + USD 20.11.

I haven't looked into the details yet but do these tests pass for you on Windows? 

```
C:\Users\sabrih\workspace\build\RelWithDebInfo>ctest -L mtoh
Test project C:/Users/sabrih/workspace/build/RelWithDebInfo
    Start 116: testMtohBasicRender
1/4 Test #116: testMtohBasicRender ..............***Failed   15.63 sec
    Start 117: testMtohCommand
2/4 Test #117: testMtohCommand ..................***Failed   20.12 sec
    Start 118: testMtohDagChanges
3/4 Test #118: testMtohDagChanges ...............***Failed   23.46 sec
    Start 119: testMtohVisibility
4/4 Test #119: testMtohVisibility ...............***Failed   22.46 sec

0% tests passed, 4 tests failed out of 4

Label Time Summary:
interactive    =  81.67 sec*proc (4 tests)
mtoh           =  81.67 sec*proc (4 tests)

Total Test time (real) =  82.70 sec

The following tests FAILED:
        116 - testMtohBasicRender (Failed)
        117 - testMtohCommand (Failed)
        118 - testMtohDagChanges (Failed)
        119 - testMtohVisibility (Failed)
Errors while running CTest
```

pmolodo commented:
> Hi @pmolodo, all 4 tests fail for me on Windows. I ran these test against Maya PR 120 + USD 20.11.
> 
> I haven't looked into the details yet but do these tests pass for you on Windows?

Ah - looks like they were only passing in our Luma environment (which was setting some env vars on it's own).  Fixed by adding mtoh to the MAYA_PLUG_IN_PATH for tests:

https://github.com/Autodesk/maya-usd/pull/915/commits/a3ebb650554e709c3bbfc5bc59f683868a5bacaf

pmolodo commented on test/lib/mtoh/testMtohCommand.py:
```diff
@@ -0,0 +1,107 @@
+        renderers = cmds.mtoh(listRenderers=1)
+        self.assertEqual(renderers, cmds.mtoh(lr=1))
+        self.assertIn(mtohUtils.HD_STORM, renderers)
+        self.assertIn("HdEmbreeRendererPlugin", renderers)
```
Should be fixed in: https://github.com/Autodesk/maya-usd/pull/915/commits/fd5fe3ca8da8eccbefef52ecd640fb34a6f8c896

pmolodo commented on test/lib/mtoh/testMtohCommand.py:
```diff
@@ -0,0 +1,107 @@
+
+        activeRenderers = cmds.mtoh(listActiveRenderers=1)
+        self.assertEqual(activeRenderers, cmds.mtoh(lar=1))
+        self.assertEqual(activeRenderers, ["HdEmbreeRendererPlugin"])
```
Should be fixed in: https://github.com/Autodesk/maya-usd/pull/915/commits/fd5fe3ca8da8eccbefef52ecd640fb34a6f8c896

pmolodo commented on test/lib/mtoh/testMtohCommand.py:
```diff
@@ -0,0 +1,107 @@
+                                getRendererDisplayName=True)
+        self.assertEqual(displayName, cmds.mtoh(r="HdEmbreeRendererPlugin",
+                                                gn=True))
+        self.assertEqual(displayName, "Embree")
```
Should be fixed in: https://github.com/Autodesk/maya-usd/pull/915/commits/fd5fe3ca8da8eccbefef52ecd640fb34a6f8c896

pmolodo commented on test/lib/mtoh/testMtohBasicRender.py:
```diff
@@ -0,0 +1,75 @@
+        self.assertSnapshotEqual("flat_orange.png")
+        self.assertRaises(AssertionError,
+                          self.assertSnapshotEqual, "flat_orange_bad.png")
+        self.assertSnapshotClose("flat_orange_bad.png", 17515 / 163200000.0)
```
Good point, added an explanation: https://github.com/Autodesk/maya-usd/pull/915/commits/e619f1e4f1cf80ce4402b53a096835a31291d7fe

pmolodo commented:
Ok, think I've addressed all the notes - I also made two other small changes:

- made the building of the mtoh tests only happen if building mtoh (https://github.com/Autodesk/maya-usd/pull/915/commits/06f439660fe7866abf52c70e3fed938c6e4f0632)
- made interactive tests output to sys.__stderr__ by default, which makes their results visible without having to run an interactive maya (https://github.com/Autodesk/maya-usd/pull/915/commits/475754d656509a9114af0735cf5e14800074b3ca)

HamedSabri-adsk commented:
@pmolodo  testMtohCommand  is now passing for me with your new changes, however the rest of the mtoh tests are still failing.
```
C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo>ctest -L mtoh
Test project C:/Users/sabrih/Desktop/workspace/build/RelWithDebInfo
    Start 95: testMtohBasicRender
1/4 Test #95: testMtohBasicRender ..............***Failed   19.24 sec
    Start 96: testMtohCommand
2/4 Test #96: testMtohCommand ..................   Passed   15.97 sec
    Start 97: testMtohDagChanges
3/4 Test #97: testMtohDagChanges ...............***Failed   18.02 sec
    Start 98: testMtohVisibility
4/4 Test #98: testMtohVisibility ...............***Failed   18.63 sec

25% tests passed, 3 tests failed out of 4
```
I also can't still see the test output in console. I am running them with -VV. Do I need to do something differently to see them? 

I picked testMtohBasicRender and set verbosity to 2 for the TextTestRunner

`unittest.TextTestRunner(stream=sys.__stderr__, verbosity=2)`

which then gave me a little bit more information where the error is coming.

`test_flat_orange (__main__.TestSnapshot) ... ERROR`

I then started to manually comment out chunk of codes to see where things go south. In this case, it happens when we hit 

`self.assertSnapshotEqual("flat_orange.png")`

pmolodo commented:
Hmm - using `-V` should be all you need to see output.  For instance, I invoke with `ctest -V -R testMtohBasicRender`, and if it succeeds, I will see:

```
114: ..
114: ----------------------------------------------------------------------
114: Ran 2 tests in 3.340s
114:
114: OK
```

...whereas if it has an unexpected error, I get a traceback:

```
114: E.
114: ======================================================================
114: ERROR: test_flat_orange (__main__.TestSnapshot)
114: ----------------------------------------------------------------------
114: Traceback (most recent call last):
114:   File "D:/Projects/Dev/USD/maya-usd/work/build/Release/test/lib/mtoh\testMtohBasicRender.py", line 50, in test_flat_orange
114:     1/0
114: ZeroDivisionError: integer division or modulo by zero
114:
114: ----------------------------------------------------------------------
114: Ran 2 tests in 2.044s
114:
114: FAILED (errors=1)
```

That's without making any further changes to the `TextTestRunner` call.

Is maya crashing? That might explain the lack of a traceback... and possibly the lack of ouput (ie, if the stream wasn't flushed before crashing).

I'm not able to replicate - but maybe it's a python 2/3 thing? I was testing with python-2 - which were you using?

HamedSabri-adsk commented:
Maya doesn't crash for me. 

> but maybe it's a python 2/3 thing

Good point. I am on Python 3 only these days and could be the case here.  Can you try with Python-3 and see if you are able to reproduce it?

pmolodo commented:
K, I'll trying building a python-3 USD.  Haven't done that on Windows yet, but I was going to need to sooner or later anyway...

pmolodo commented:
So, got a windows python-3 build, and I now see what you do - it errors, and doesn't even give any output.  Odd.

pmolodo commented:
So, I fixed the output to the terminal, which allowed me to quickly solve the python-3 issue.  (The terminal output problem was due to streams not being flushed... and the python-3 issue was due to use of xrange!)

I'd actually like to apply the usage of `fixturesUtils.runTests(globals())`, seen [here](https://github.com/Autodesk/maya-usd/pull/915/commits/c6cc95c9661c92ef6ee8578100934707b21398c1#diff-8faeaa88b92aa26a5ed0c3a2204ae96d1803bef7dc64f00664b0ea13c14e2c70R61), to all the interactive maya tests; but this PR was already starting to get bloated, so I'll hold that for a separate PR.  (If you want, I can also pull out all the generic-test changes already in this PR - ie, c6cc95c9661c92ef6ee8578100934707b21398c1, 475754d656509a9114af0735cf5e14800074b3ca, 5a503a3b8fcf3ebddb2e3cf7664f9e69da50f0dc, 2dcf1c47f1b215e024f0e31b8714a6fc9f8cbdc0.  These changes aren't strictly necessary for the mtoh tests, though they were very useful when developing and to pinpoint problems!)

HamedSabri-adsk commented:
Thank you for the fix. 3 of the tests now pass but testMtohVisibility failed for me.

```
C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo>ctest -L mtoh
Test project C:/Users/sabrih/Desktop/workspace/build/RelWithDebInfo
    Start 121: testMtohBasicRender
1/4 Test #121: testMtohBasicRender ..............   Passed   21.85 sec
    Start 122: testMtohCommand
2/4 Test #122: testMtohCommand ..................   Passed   17.36 sec
    Start 123: testMtohDagChanges
3/4 Test #123: testMtohDagChanges ...............   Passed   25.55 sec
    Start 124: testMtohVisibility
4/4 Test #124: testMtohVisibility ...............***Failed   20.58 sec
```
I ran testMtohVisibility with both -V and -VV and still can't get any information on the terminal output :(

```
ctest -R testMtohVisibility -V

124: ......E
1/1 Test #124: testMtohVisibility ...............***Failed   16.00 sec

0% tests passed, 1 tests failed out of 1
```


> I'd actually like to apply the usage of fixturesUtils.runTests(globals()), seen here, to all the interactive maya tests;

That's a very good change but let's have it in a separate PR.

> I can also pull out all the generic-test changes already in this PR

Should be fine to have them in this PR. Thanks

pmolodo commented:
Odd - all the tests now pass for me.

As for the lack of output - that's probably because I only applied the use of `fixturesUtils.runTests(globals())` to the one test that was failing for me.  I pushed a new commit where I apply it to all the mtoh tests.

Does it give you a traceback with that?

HamedSabri-adsk commented:
ok, now I can see the output which is good.

So somehow this test didn't generate valid images ( "cube_unselected.png", "nothing.png ) which is odd. I am looking at the other test ( MtohBasicRender ) and the same images got generated fine there.

See attached zip file for these 2 images.

[MtohVisibility_images.zip](https://github.com/Autodesk/maya-usd/files/5640260/MtohVisibility_images.zip)

![image](https://user-images.githubusercontent.com/48299423/101114872-5ca89380-35b0-11eb-8145-e8accefcb885.png)


```
24: ======================================================================
124: ERROR: test_toggleTransVis (__main__.TestCommand)
124: ----------------------------------------------------------------------
124: Traceback (most recent call last):
124:   File "C:/Users/sabrih/Desktop/workspace/build/RelWithDebInfo/test/lib/mtoh\testMtohVisibility.py", line 26, in test_toggleTransVis
124:     self.assertSnapshotClose("cube_unselected.png")
124:   File "C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\python\mtohUtils.py", line 44, in assertSnapshotClose
124:     refImage, maxAvgChannelDiff=maxAvgChannelDiff)
124:   File "C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\python\imageUtils.py", line 167, in assertSnapshotClose
124:     maxAvgChannelDiff=maxAvgChannelDiff)
124:   File "C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\python\imageUtils.py", line 149, in assertImagesClose
124:     diff = imageDiff(imagePath1, imagePath2)
124:   File "C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\python\imageUtils.py", line 91, in imageDiff
124:     img1 = loadImage(imagePath1)
124:   File "C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\python\imageUtils.py", line 88, in loadImage
124:     raise ValueError("unable to load image: {}".format(imagePath))
124: ValueError: unable to load image: C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo\test\lib\mtoh\MtohVisibility\cube_unselected.png
124:
124: ----------------------------------------------------------------------
124: Ran 7 tests in 4.026s
124:
124: FAILED (errors=1)
1/1 Test #124: testMtohVisibility ...............***Failed   18.39 sec
```

pmolodo commented:
Ah - so realized what's causing problems is the reference images for that test are actually symlinked.  Worked for me because I have perms to make symlinks, but I think that's not true by default on windows.

Pushed a change removing the symlinks.

Now, just need to figure out what's causing the preflight failures...

HamedSabri-adsk commented:
All tests now pass for me.  Going with "dots" by pointing at those image resources in build dir is a better option. Yes, using mklink on Windows requires admin permission. 

```
C:\Users\sabrih\Desktop\workspace\build\RelWithDebInfo>ctest -j 48 -L mtoh
Test project C:/Users/sabrih/Desktop/workspace/build/RelWithDebInfo
    Start 125: testMtohVisibility
    Start 122: testMtohBasicRender
    Start 123: testMtohCommand
    Start 124: testMtohDagChanges
1/4 Test #123: testMtohCommand ..................   Passed   20.81 sec
2/4 Test #124: testMtohDagChanges ...............   Passed   24.67 sec
3/4 Test #122: testMtohBasicRender ..............   Passed   24.88 sec
4/4 Test #125: testMtohVisibility ...............   Passed   25.12 sec
```

HamedSabri-adsk reviewed (approved):
Looking good.

pmolodo commented:
So, removing the setting of MAYA_APP_DIR makes the tests not timeout.  There's something about the preflight build environment that for some reason makes the setup of the settings directory very slow - looking at several runs, which tests were timing out wasn't consistent, so I don't think it's anything specific to any test.

@kxl-adsk Do you think you could ask the dev ops people for thoughts on why this might be the case? For now, we can merge this without this... though it means that each user's personal settings are being used in tests, which could make the image-diff tests somewhat fragile.

kxl-adsk commented:
@pmolodo I was checking this PR prior to merge and the following tests failed
```
        102 - testMtohBasicRender (Failed)
        104 - testMtohDagChanges (Failed)
        105 - testMtohVisibility (Failed)
```

I'm not sure if it has to do with the build configuration (above was building Maya 2020 with USD 19.11) or my build of USD. Are you seeing this coding error when running the tests?
```
Coding Error: in Read at line 133 of \usd\pxr\imaging\lib\glf\uvtexturestoragedata.cpp -- Unsupported texture storage data type
```

Finally, we want the test folder structure to follow the lib folder structure. There is some cleanup already to do, and this PR is (almost) ready to go in, but maybe after it's merged you could adjust the structure by moving `test/lib/mtoh` -> `test/lib/mayaUsd/render/mayaToHydra`?

kxl-adsk commented:
OK, so tests are failing because of USD 19.11 (the version of Maya, or platform is irrelevant). Starting USD 20.02, everything is passing. FWIW. I had #961 merged as well, but not sure if it's related. 

If we have an explanation for the failure, let's disable the tests for USD 19.11 and document why we don't expect them to pass.

pmolodo commented:
I don't know offhand why it's not working with 19.11 - I would expect it to. I'll look at it...

pmolodo commented:
So... finally got a 19.11 + 2020 build, and confirmed the failure.  Basically, it just looks like either some sort of colorspace or lighting falloff issue - the generated image in 19.11 basically looks correct, except it's darker.  I could try and track down exact what changed and why, but I suspect it will boil down to finding a USD commit message like "we fixed colorspaces for ZZ", or "we changed the default lighting falloff to **2 for the default light".

Anyway, if it works in all other usd versions, whatever the change was it seems to have stablized, so I say we just add the skip as you suggested... I'll make that change and push it.

pmolodo commented:
...or do you think we need to give a more precise reason for failure?

kxl-adsk commented:
I can agree with your conclusion. Given we have things covered for latest and future releases of USD, I wouldn’t invest more in debugging it. 

I’m also happy we are now covered, even if these are just basic tests.

pmolodo commented:
Made the tests only run if USD > 19.11, and also moved the test folder as requested.  (Also renamed the test resource folders, by appending 'Test', to make them match conventions in other mayaUsd tests.)

pmolodo commented:
Tests failed again - this time due to `testUsdExportPackage` failing to delete a file, looks like. I'm going to rebase off latest dev, and see if that helps anything...
